[2D] Enable 2D FSDP+TP model.load_state_dict()

wz337 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110925
* #110846
* #110831

This PR adds a all_gather_dtensor() method to fsdp/_fsdp_extensions.py and the actual implementation in tensor/parallel/fsdp.py. This enables FSDP to load 2D DTensor state_dict into model when calling `model.load_state_dict()`.

cc. @fegin

wz337 committed:
enable 2D FSDP+TP load_state_dict()

[ghstack-poisoned]

wz337 committed:
Update on "[2D] Enable 2D FSDP+TP model.load_state_dict()"


This PR adds a all_gather_dtensor() method to fsdp/_fsdp_extensions.py and the actual implementation in tensor/parallel/fsdp.py. This enables FSDP to load 2D DTensor state_dict into model when calling `model.load_state_dict()`.

cc. fegin 


[ghstack-poisoned]

fduwjj commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -190,3 +190,24 @@ def _create_chunk_dtensor(
     )
+
+
+def _all_gather_dtensor(
```
Can we merge this logic and the one in `torch/distributed/tensor/parallel/fsdp.py` into one common place?

wz337 replied:
We cannot. This `_all_gather_dtensor` is internal to FSDP.  We are following the extension design here. See this function as an example: https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_fsdp_extensions.py#L91 

Essentially, there are two code paths:
FSDP only --> _all_gather_dtensor()
FSDP + TP -> _extensions.all_gather_dtensor

wz337 committed:
Update on "[2D] Enable 2D FSDP+TP model.load_state_dict()"


This PR adds a all_gather_dtensor() method to fsdp/_fsdp_extensions.py and the actual implementation in tensor/parallel/fsdp.py. This enables FSDP to load 2D DTensor state_dict into model when calling `model.load_state_dict()`.

cc. fegin 


[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -190,3 +190,24 @@ def _create_chunk_dtensor(
+
+def _all_gather_dtensor(
+    tensor: DTensor,
+    parent_mesh: DeviceMesh,
```
This typing should be optional.

wz337 committed:
Update on "[2D] Enable 2D FSDP+TP model.load_state_dict()"


This PR adds a all_gather_dtensor() method to fsdp/_fsdp_extensions.py and the actual implementation in tensor/parallel/fsdp.py. This enables FSDP to load 2D DTensor state_dict into model when calling `model.load_state_dict()`.

cc. fegin 


[ghstack-poisoned]

wz337 committed:
Update on "[2D] Enable 2D FSDP+TP model.load_state_dict()"


This PR adds a all_gather_dtensor() method to fsdp/_fsdp_extensions.py and the actual implementation in tensor/parallel/fsdp.py. This enables FSDP to load 2D DTensor state_dict into model when calling `model.load_state_dict()`.

cc. fegin 


[ghstack-poisoned]

fegin reviewed (approved):
LGTM

wz337 commented:
@pytorchmergebot merge
